Pal 866 policy service audit error handling

m78233 opened the pull request:
Previous PR was failing on SonarQube test for a different PR.  Created a new PR after merging.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/gchq/Palisade-services?pullRequest=219) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/gchq/Palisade-services?pullRequest=219) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: ac74475<br/>:white_check_mark: pd104923<br/>:white_check_mark: developer01189998819991197253<br/>:x: m78233<br/><hr/>**m78233** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/gchq/Palisade-services?pullRequest=219) it.</sub>

Commit:
PAL-866_Policy_Service_Audit_Error_Handling: Squashed existing check-ins as I had a mix of signed and unsigned commits.  This commit is all signed and consists of the changes that are related to:
Changes for the ticket
Added:
	AuditablePolicyRecordResponse
	AuditablePolicyResourceResponse
	AuditablePolicyRecordResponseTest
	AuditablePolicyResourceResponseTest
	KafkaHealthIndicator
	PolicyServiceAsyncProxyTest

Changed:
AuditErrorMessageTest
PolicyResponseTest
AuditErrorMessage
PolicyRequest
PolicyResponse
PolicyServiceAsyncProxy
PolicyServiceCachingProxy
SerDesConfig
AkkaComponentsConfig
AkkaRunnableGraph
application.yaml
application-aka.yaml
ApplicationTestData

2) Changes based on the initial feedback from the code review.

Signed-off-by: m78233 <m78233@g61839550+m78233@users.noreply.github.com>

Commit:
Merge remote-tracking branch 'origin/develop' into PAL-866_Policy_Service_Audit_Error_Handling

pd104923 committed:
Merge branch 'develop' into PAL-866_Policy_Service_Audit_Error_Handling

pd104923 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/config/AkkaRunnableGraph.java:
```diff
@@ -70,68 +72,64 @@ KafkaProducerService kafkaProducerService(
+                // Build producer message, copying the partition, keeping track of original message
+                .map((Pair<CommittableMessage<String, PolicyRequest>, AuditablePolicyRecordResponse> messageAndResponse) -> {
+                    ConsumerRecord<String, PolicyRequest> requestRecord = messageAndResponse.first().record();
+                    Committable commitable = (Committable) messageAndResponse.first().committableOffset();
```
```suggestion
                    Committable committable = messageAndResponse.first().committableOffset();
```

m78233 replied:
Remove the casting  and fixed the spelling mistake.

pd104923 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/config/AkkaRunnableGraph.java:
```diff
@@ -70,68 +72,64 @@ KafkaProducerService kafkaProducerService(
+                            .map(audit -> ProducerMessage.single(
+                                    new ProducerRecord<>(errorTopic.getName(), requestRecord.partition(), requestRecord.key(),
+                                            SerDesConfig.errorValueSerializer().serialize(null, audit), requestRecord.headers()),
+                                    (Committable) messageAndResponse.first().committableOffset()))
```
Can we not use the committable variable here?
```suggestion
                                    committable))
```

m78233 replied:
Did the change and corrected the spelling

pd104923 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/config/AkkaRunnableGraph.java:
```diff
@@ -70,68 +72,64 @@ KafkaProducerService kafkaProducerService(
+                            .orElse(ProducerMessage.single(
+                                    new ProducerRecord<>(outputTopic.getName(), requestRecord.partition(), requestRecord.key(),
+                                            SerDesConfig.ruleValueSerializer().serialize(null, messageAndResponse.second().getPolicyResponse()), requestRecord.headers()),
+                                    commitable));
```
```suggestion
                                    committable));
```

m78233 replied:
done

pd104923 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Override
+    public Health getHealth(final boolean includeDetails) {
+        return Optional.of(performCheck())
+                .filter(healthy -> healthy)
+                .map(up -> Health.up().withDetail("group", this.groupId).build())
+                .orElseGet(() -> Health.down().withDetail("group", this.groupId).build());
```
This optional is never empty. Maybe do this:
```
if (performCheck()) {
  return Health.up().with...
} else {
  return Health.down().with...
}
```
This can also be done for the `health()` method

m78233 replied:
ok, change made

pd104923 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,102 @@
+ * Kafka health indicator. Check that the consumer group can be accessed and is registered with the cluster,
+ * if not mark the service as unhealthy.
+ */
+@Component("kafka")
+@ConditionalOnEnabledHealthIndicator("kafka")
```
This has been done differently in the User-service.
These annotations have been removed and a bean has been created in the ApplicationConfirguration class

m78233 replied:
This is setting-up the KafkaHealthIndicator as a service.  Have a look at the attribute-masking-service and note the use of the @Component annotation for the class.

pd104923 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,102 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaHealthIndicator.class);
+
+    @Value("${akka.kafka.consumer.kafka-clients.group.id}")
+    private String groupId;
```
Again, this has been done slightly differently in the User-service. The ConsumerTopicConfiguration has been added as a constructor parameter and the `groupId` value has been set in the `performCheck()` method.

m78233 replied:
See previous comment.

pd104923 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/model/PolicyRequest.java:
```diff
@@ -78,7 +78,7 @@ public String getResourceId() {
 
     @Generated
-    public Context getContext() {
+    public Context getContext()  {
```
No need for this extra space

m78233 replied:
Done

pd104923 commented on policy-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/policy/rest/HealthActuatorContractTest.java:
```diff
@@ -37,9 +38,10 @@
         classes = PolicyApplication.class,
         webEnvironment = WebEnvironment.RANDOM_PORT,
-        properties = {"spring.autoconfigure.exclude=org.springframework.boot.autoconfigure.data.redis.RedisAutoConfiguration"}
+        properties = {"management.health.kafka.enabled=false", "spring.autoconfigure.exclude=org.springframework.boot.autoconfigure.data.redis.RedisAutoConfiguration"}
 )
 @ActiveProfiles("akka-test")
+@TestPropertySource(properties = "management.health.kafka.enabled=false")
```
Why is this value in 2 places?

m78233 replied:
Too quick with the copy and paste.  Have removed the whole line starting with @TestPropertySource... plus the second property in the @SpringBootTest properties lines starting with "spring.autoconfigure..."

pd104923 commented on policy-service/src/contract-tests/java/uk/gov/gchq/palisade/contract/policy/kafka/KafkaContractTest.java:
```diff
@@ -289,6 +298,22 @@ void testNoSuchPolicyExceptionIsThrown() {
         results.removeLast();
         assertThat(results).isEmpty();
+
+        assertAll("Asserting on the error topic",
+                // One error is produced
+                () -> assertThat(errorResults)
+                        .hasSize(1),
+
+                // The error has the relevant headers, including the token
+                () -> assertThat(errorResults)
+                        .allSatisfy(result ->
+                                assertThat(result.headers().lastHeader(Token.HEADER).value())
+                                        .isEqualTo(ContractTestData.REQUEST_TOKEN.getBytes())),
+
+                // The error has a message that contains the throwable exception, and the message
+                () -> assertThat(errorResults.get(0).value().get("error").get("message").asText())
+                        .isEqualTo(NoSuchPolicyException.class.getName() + ": No Resource Rules found for the resource")
+        );
```
```suggestion
        assertThat(errorResults)
                .hasSize(1)
                .allSatisfy(result -> assertThat(result.headers().lastHeader(Token.HEADER).value())
                        .isEqualTo(ContractTestData.REQUEST_TOKEN.getBytes()))
                .allSatisfy(result -> assertThat(errorResults.get(0).value().get("error").get("message").asText())
                        .isEqualTo(NoSuchPolicyException.class.getName() + ": No Resource Rules found for the resource"));
```

m78233 replied:
I want to keep the comments.  I think it helps in reading the code.

pd104923 reviewed (changes requested):
A few minor changes and a couple left over from the previous PR but looking good.

m78233 committed:
PAL-866_Policy_Service_Audit_Error_Handling: Squashed existing check-ins as I had a mix of signed and unsigned commits.  This commit is all signed and consists of the changes that are related to:
Changes for the ticket
Added:
	AuditablePolicyRecordResponse
	AuditablePolicyResourceResponse
	AuditablePolicyRecordResponseTest
	AuditablePolicyResourceResponseTest
	KafkaHealthIndicator
	PolicyServiceAsyncProxyTest

Changed:
AuditErrorMessageTest
PolicyResponseTest
AuditErrorMessage
PolicyRequest
PolicyResponse
PolicyServiceAsyncProxy
PolicyServiceCachingProxy
SerDesConfig
AkkaComponentsConfig
AkkaRunnableGraph
application.yaml
application-aka.yaml
ApplicationTestData

2) Changes based on the initial feedback from the code review.

m78233 committed:
Merge remote-tracking branch 'origin/PAL-866_Policy_Service_Audit_Error_Handling' into PAL-866_Policy_Service_Audit_Error_Handling

m78233 committed:
PAL-866_Policy_Service_Audit_Error_Handling: changes based on second code review

Signed-off-by: m78233 <61839550+m78233@users.noreply.github.com>

m78233 committed:
PAL-866_Policy_Service_Audit_Error_Handling: code clean-up

Signed-off-by: m78233 <61839550+m78233@users.noreply.github.com>

m78233 committed:
PAL-866_Policy_Service_Audit_Error_Handling: changed Health Indicator from a Component to a Bean declared in the Application Configuration.

Signed-off-by: m78233 <61839550+m78233@users.noreply.github.com>

pd104923 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/stream/config/AkkaRunnableGraph.java:
```diff
@@ -70,68 +72,64 @@ KafkaProducerService kafkaProducerService(
             final ProducerTopicConfiguration topicConfiguration,
             final PolicyServiceAsyncProxy service) {
         // Get output topic from config
         Topic outputTopic = topicConfiguration.getTopics().get("output-topic");
+        Topic errorTopic = topicConfiguration.getTopics().get("error-topic");
 
         // Read messages from the stream source
+        //  return source
         return source
-                .map(message -> new Pair<>(message, Optional.ofNullable(message.record().value())))
+                .map(committableMessage -> new Pair<>(committableMessage, committableMessage.record().value()))
```
I think this still needs to be an Optional because of the `START` and `END` messages as these will have a request value of `null`.

m78233 replied:
The getResourceRules has been modified to handle a null.  All of the messages goes in there and, it creates a "container" object that reflects the state things with the PolicyRequest, Rules, Exception and the modified response.  Depending on the circumstances, some of these will be empty.  The original approach, some of this was being handled externally in the RunnableGraph and the the Optional was being used as part of the logic for if and when the getResourceRules was going to be called.

pd104923 reviewed (changes requested):
One small change/comment added. There's also a comment in the KafkaHealthInidcator from before that has not been dealt with.

m78233 committed:
PAL-866_Policy_Service_Audit_Error_Handling: changes to the health and getHealth methods based on comments in the review.

Signed-off-by: m78233 <61839550+m78233@users.noreply.github.com>

developer01189998819991197253 committed:
Merge branch 'develop' into PAL-866_Policy_Service_Audit_Error_Handling

dbz120896 commented on policy-service/src/component-tests/java/uk/gov/gchq/palisade/component/policy/model/AuditablePolicyRecordResponseTest.java:
```diff
@@ -0,0 +1,138 @@
+     */
+    @Test
+    void testAuditablePolicyRecordResponseSerialisingAndDeserialising() throws IOException {
+        AuditablePolicyRecordResponse auditablePolicyRecordResponse = AuditablePolicyRecordResponse.Builder.create().withPolicyResponse(RESPONSE).withNoErrors();
```
This can be simplified greatly:
```
        var expected = AuditablePolicyRecordResponse.Builder.create().withPolicyResponse(RESPONSE).withNoErrors();
        var actual = jsonTester.parse(jsonTester.write(expected).getJson()).getObject();
        assertThat(actual)
            .as("check using equals()").isEqualTo(expected)
            .usingRecursiveComparison()
            .as("check using recursion").isEqualTo(expected);
```
When testing serialisation and then desrialisation there is now need to assert individual fields. The standard `isEqual()` and it's recursive form will catch anything amiss.

m78233 replied:
This is a simple example of what I was talking about in the slack.  I appreciate the recursive comparison tells all.  The downside is when I mistype the parameter "resourceId". as  "resourceID", it will be the test that checks each parameter in the JSON Object and the Java Object have both been this done correctly and there is no small typo buried in there that will bite me.  The recursive tells me that something is wrong, but I have to work at it to what went wrong.  So when I write the code & tests, this find grained test is useful.  Once the code is right, the recursive is more useful as then all we want to know is did some other change break this code.   I think there is another plus is that a new programmer should be able to go to the tests and be able to understand what does the class do.  Here I think I would gravitate to the "overkill" version as it spells it out in detail.

kg981167224 replied:
There's pluses and minuses for either way. Really hate the use of vars though,

dbz120896 commented on policy-service/src/component-tests/java/uk/gov/gchq/palisade/component/policy/model/AuditablePolicyRecordResponseTest.java:
```diff
@@ -0,0 +1,138 @@
+     */
+    @Test
+    void testAuditablePolicyRecordResponseExceptionSerialisingAndDeserialising() throws IOException {
+        AuditablePolicyRecordResponse auditablePolicyRecordResponse = AuditablePolicyRecordResponse.Builder.create().withPolicyResponse(null).withAuditErrorMessage(AUDIT_ERROR_MESSAGE);
```
As above, this can be changed as follows:
```
        var expected = AuditablePolicyRecordResponse.Builder.create().withPolicyResponse(RESPONSE).withAuditErrorMessage(AUDIT_ERROR_MESSAGE);
        var actual = jsonTester.parse(jsonTester.write(expected).getJson()).getObject();
        assertThat(actual)
            .as("check using equals()").isEqualTo(expected)
            .usingRecursiveComparison()
            .as("check using recursion").isEqualTo(expected);
```
Any serialisation test can use just these 4 lines.

m78233 replied:
I have added  the missing equals() and hashCode() to the class.  I have included both the recursive comparison and the isEqualTo to each of the tests. I have started to add the self documenting  .as(text) to the methods in the test class will adopt that going forward.

I like the minimalism that you have suggested for this test, but I am going to hold off on deleting the existing assertions for the moment.  I was asked to put in the finer detailed assertions for each attribute in a code review that was previously on the classes in the model package and this became convention.  We want to change this, I feel that we need to run it up the flag pole.

dbz120896 replied:
Yeah, this bit of code should be discussed. I think the reason for doing it is so that if the equality check failes, the test highlights which field. But by using `.usingRecursiveComparison()` and then doing an equals check, AssertJ will indicate in the test which fields did not match. The good thing here is that if another field is added, it will automatically be included in the test :)

dbz120896 commented on policy-service/src/component-tests/java/uk/gov/gchq/palisade/component/policy/model/AuditablePolicyResourceResponseTest.java:
```diff
@@ -0,0 +1,169 @@
+     */
+    @Test
+    void testInitalAuditablePolicyResourceResponseSerialisingAndDeserialising() throws IOException {
+        AuditablePolicyResourceResponse auditablePolicyResourceResponse = AuditablePolicyResourceResponse.Builder.create()
```
Use
```
        var expected = AuditablePolicyResourceResponse.Builder.create()
            .withPolicyRequest(REQUEST)
            .withRules(RULES)
            .withNoErrors()
            .withNoNoModifiedResponse();

        var actual = jsonTester.parse(jsonTester.write(expected).getJson()).getObject();

        assertThat(actual)
            .as("check using equals()").isEqualTo(expected)
            .usingRecursiveComparison()
            .as("check using recursion").isEqualTo(expected);
```
instead. This code caught the fact that `AuditablePolicyResourceResponse` does not have `equals()` and `hashcode()` implemented.

m78233 replied:
Added the missing equal and hashCode method to the class.  For the rest of the discussion see the previous comment.

dbz120896 commented on policy-service/src/component-tests/java/uk/gov/gchq/palisade/component/policy/model/AuditablePolicyResourceResponseTest.java:
```diff
@@ -0,0 +1,169 @@
+     */
+    @Test
+    void testModifiedAuditablePolicyResourceResponseSerialisingAndDeserialising() throws IOException {
+        AuditablePolicyResourceResponse auditablePolicyResourceResponse = AuditablePolicyResourceResponse.Builder.create()
```
Use
```
        var expected = AuditablePolicyResourceResponse.Builder.create()
            .withPolicyRequest(REQUEST)
            .withRules(RULES)
            .withNoErrors()
            .withModifiedResource(LEAF_RESOURCE);

        var actual = jsonTester.parse(jsonTester.write(expected).getJson()).getObject();

        assertThat(actual)
            .as("check using equals()").isEqualTo(expected)
            .usingRecursiveComparison()
            .as("check using recursion").isEqualTo(expected);
```
instead. This code caught the fact that `AuditablePolicyResourceResponse` does not have `equals()` and `hashcode()` implemented.

m78233 replied:
See previous comment

dbz120896 commented on policy-service/src/component-tests/java/uk/gov/gchq/palisade/component/policy/model/AuditablePolicyResourceResponseTest.java:
```diff
@@ -0,0 +1,169 @@
+     */
+    @Test
+    void testErrorAuditablePolicyResourceResponseSerialisingAndDeserialising() throws IOException {
+        AuditablePolicyResourceResponse auditablePolicyResourceResponse = AuditablePolicyResourceResponse.Builder.create()
```
Use
```
        var expected = AuditablePolicyResourceResponse.Builder.create()
            .withPolicyRequest(REQUEST)
            .withRules(RULES)
            .withAuditErrorMessage(AUDIT_ERROR_MESSAGE)
            .withModifiedResource(LEAF_RESOURCE);

        var actual = jsonTester.parse(jsonTester.write(expected).getJson()).getObject();

        assertThat(actual)
            .as("check using equals()").isEqualTo(expected)
            .usingRecursiveComparison()
            .as("check using recursion").isEqualTo(expected);
```
instead. This code caught the fact that `AuditablePolicyResourceResponse` does not have `equals()` and `hashcode()` implemented.

m78233 replied:
See previous comment.

dbz120896 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,103 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(KafkaHealthIndicator.class);
+
+    private String groupId;
```
Remove `groupId` for thread safety.

m78233 replied:
This class is a copy and paste from the work done in the attribute-masking-service.  For whatever reason, the annotation  @Value("${akka.kafka.consumer.kafka-clients.group.id}") is not showing in the class.

dbz120896 commented on policy-service/src/main/java/uk/gov/gchq/palisade/service/policy/service/KafkaHealthIndicator.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    private boolean performCheck() {
+        groupId = topicConfiguration.getKafkaClients().get("group.id");
```
Extract `group.id` to a constant.

m78233 replied:
See previous comment.

m78233 committed:
PAL-866_Policy_Service: changes based on feedback in code review

Signed-off-by: m78233 <61839550+m78233@users.noreply.github.com>

m78233 committed:
PAL-866_Policy_Service: changes based on feedback in code review

Signed-off-by: m78233 <61839550+m78233@users.noreply.github.com>

m78233 committed:
PAL-866_Policy_Service: changes for the HealthActuatorContract based on feedback in code review

Signed-off-by: m78233 <61839550+m78233@users.noreply.github.com>

developer01189998819991197253 committed:
Merge branch 'develop' into PAL-866_Policy_Service_Audit_Error_Handling

m78233 committed:
PAL-866_Policy_Service: removed unused methods

Signed-off-by: m78233 <61839550+m78233@users.noreply.github.com>

m78233 committed:
PAL-866_Policy_Service: KafkaHealthIndicator removed unused variable and made instance variable final

Signed-off-by: m78233 <61839550+m78233@users.noreply.github.com>

ac74475 committed:
Merge branch 'develop' into PAL-866_Policy_Service_Audit_Error_Handling

m78233 committed:
PAL-866_Policy_Service: code clean-up

Signed-off-by: m78233 <61839550+m78233@users.noreply.github.com>

m78233 committed:
Merge remote-tracking branch 'origin/PAL-866_Policy_Service_Audit_Error_Handling' into PAL-866_Policy_Service_Audit_Error_Handling

dev930018 commented:
Closed because of bad CLA, see https://github.com/gchq/Palisade-services/pull/225
